### PR TITLE
*: do not report error if objectID is invalid

### LINKED ISSUE #22950: Do not report error for placement rules that has an non-exist partition ID
## Development Task

It is a known issue. Placement rules are cached in memory, persisted in PD. `drop table` should drop placement rules attached to the table. But you may recover the table later and expect that placement rules will recover, too.

We solve that by delaying the deletion of placement rules, check #20575. But as we don't really drop the persisted rules in PD. TiDB will reload the deleted rules back when:

1. restart
2. DDL infoschema fullLoad/snapshot load

Then you met the error `mysql root@127.0.0.1:(none)> select * from information_schema.placement_policy
(1105, "Can't find partition by id 59")`

## The full reproduce instructions

```
# start TiUP and add some rules

mysql root@127.0.0.1:(none)> select * from information_schema.placement_policy
+-------------+-------------+---------+-------------+------------+----------------+------------+--------+----------+-------------------+
| GROUP_ID    | GROUP_INDEX | RULE_ID | SCHEMA_NAME | TABLE_NAME | PARTITION_NAME | INDEX_NAME | ROLE   | REPLICAS | CONSTRAINTS       |
+-------------+-------------+---------+-------------+------------+----------------+------------+--------+----------+-------------------+
| TiDB_DDL_59 | 3           | leader  | test        | t          | p0             | <null>     | leader | 1        | "-engine=tiflash" |
| TiDB_DDL_59 | 3           | voter   | test        | t          | p0             | <null>     | voter  | 3        | "-engine=tiflash" |
+-------------+-------------+---------+-------------+------------+----------------+------------+--------+----------+-------------------+
mysql root@127.0.0.1:(none)> drop table test.t
mysql root@127.0.0.1:(none)> select * from information_schema.placement_policy
+----------+-------------+---------+-------------+------------+----------------+------------+------+----------+-------------+
| GROUP_ID | GROUP_INDEX | RULE_ID | SCHEMA_NAME | TABLE_NAME | PARTITION_NAME | INDEX_NAME | ROLE | REPLICAS | CONSTRAINTS |
+----------+-------------+---------+-------------+------------+----------------+------------+------+----------+-------------+

# restart(scale-in + scale-out) the TiDB node

mysql root@127.0.0.1:(none)> select * from information_schema.placement_policy
(1105, "Can't find partition by id 59")
```

### PR DESCRIPTION
**Pull request by xhebox**
Signed-off-by: xhe <xw897002528@gmail.com>

### What problem does this PR solve?

Issue Number: close #22950

Problem Summary: Refer to the issue.

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

See the issues for reproducing instructions.

### Release note 

- No release note

### CODE REVIEW

**Comment by xhebox** [comment]
cc @djshow832

File: executor/infoschema_reader.go [hunk 1]
```diff
@@ -1848,12 +1848,15 @@ func (e *memtableRetriever) setDataForPlacementPolicy(ctx sessionctx.Context) er
...
 		}
 		// Currently, only partitions have placement rules.
+		var tbName, dbName, ptName string
 		tb, db, part := is.FindTableByPartitionID(id)
-		if tb == nil {
-			return errors.Errorf("Can't find partition by id %d", id)
-		}
-		if checker != nil && !checker.RequestVerification(ctx.GetSessionVars().ActiveRoles, db.Name.L, tb.Meta().Name.L, "", mysql.SelectPriv) {
-			continue
+		if tb != nil {
```
Comment (djshow832) [inline executor/infoschema_reader.go:1853]: Still show the row even if the partition doesn't exist?

Comment (xhebox) [reply executor/infoschema_reader.go:1853]: It is intended for testing. We don't want a filtered list of rules if we do further tests. Let me add a failpoint.

File: executor/infoschema_reader.go [hunk 2]
```diff
@@ -1848,11 +1849,24 @@ func (e *memtableRetriever) setDataForPlacementPolicy(ctx sessionctx.Context) er
 			continue
 		}
 		// Currently, only partitions have placement rules.
+		var tbName, dbName, ptName string
+		skip := false
```
Comment (djshow832) [inline executor/infoschema_reader.go:1853]: Maybe `true` by default?

Comment (xhebox) [reply executor/infoschema_reader.go]: Whoops, yes. my thoughts did not flow smoothly today.

**Review by djshow832 — COMMENTED** [review]
Add a test case.
Maybe: mock an invalid partition ID and adds it into the cache, then query the system table.

File: executor/infoschema_reader.go [hunk 3]
```diff
@@ -1853,11 +1854,23 @@ func (e *memtableRetriever) setDataForPlacementPolicy(ctx sessionctx.Context) er
 			continue
 		}
 		// Currently, only partitions have placement rules.
+		var tbName, dbName, ptName string
+		skip := true
 		tb, db, part := is.FindTableByPartitionID(id)
-		if tb == nil {
-			return errors.Errorf("Can't find partition by id %d", id)
+		if tb != nil {
+			if checker == nil || checker.RequestVerification(ctx.GetSessionVars().ActiveRoles, db.Name.L, tb.Meta().Name.L, "", mysql.SelectPriv) {
```
Comment (zimulala) [inline executor/infoschema_reader.go:1861]:
```suggestion
			if tb != nil && (checker == nil || checker.RequestVerification(ctx.GetSessionVars().ActiveRoles, db.Name.L, tb.Meta().Name.L, "", mysql.SelectPriv)) {
```

File: executor/infoschema_reader.go [hunk 4]
```diff
@@ -1853,11 +1854,23 @@ func (e *memtableRetriever) setDataForPlacementPolicy(ctx sessionctx.Context) er
 			continue
 		}
 		// Currently, only partitions have placement rules.
+		var tbName, dbName, ptName string
+		skip := true
 		tb, db, part := is.FindTableByPartitionID(id)
-		if tb == nil {
-			return errors.Errorf("Can't find partition by id %d", id)
+		if tb != nil {
```
Comment (zimulala) [inline executor/infoschema_reader.go:1860]: Could we remove this `if`？

Comment (xhebox) [reply executor/infoschema_reader.go]: Done.

**Pull request merged**